pkg/executor/grant.go

```diff
@@ -790,28 +839,27 @@ func getRowsAndFields(sctx sessionctx.Context, rs sqlexec.RecordSet) ([]chunk.Ro
 	if rs == nil {
 		return nil, nil, errors.Errorf("nil recordset")
 	}
-	rows, err := getRowFromRecordSet(ctx, sctx, rs)
-	if err != nil {
-		return nil, nil, err
-	}
-	if err = rs.Close(); err != nil {
-		return nil, nil, err
-	}
-	return rows, rs.Fields(), nil
-}
-
-func getRowFromRecordSet(ctx context.Context, se sessionctx.Context, rs sqlexec.RecordSet) ([]chunk.Row, error) {
-	var rows []chunk.Row
+	var (
+		rows []chunk.Row
+		err  error
+	)
 	req := rs.NewChunk(nil)
 	for {
-		err := rs.Next(ctx, req)
+		err = rs.Next(ctx, req)
 		if err != nil || req.NumRows() == 0 {
-			return rows, err
+			break
 		}
 		iter := chunk.NewIterator4Chunk(req)
 		for r := iter.Begin(); r != iter.End(); r = iter.Next() {
 			rows = append(rows, r)
 		}
-		req = chunk.Renew(req, se.GetSessionVars().MaxChunkSize)
+		req = chunk.Renew(req, sctx.GetSessionVars().MaxChunkSize)
```

Comment:
If the table contains 1M record, it may use a lot of memory and make tidb jitter.
I once use `rows, fs, err := sctx.ExecRestrictedSQL(ctx, nil, sql)` but change it back and add comment here
https://github.com/pingcap/tidb/pull/59487/files#diff-c3393c2c85014927cf2d80a965fb61f15fdc3f375d9deefeb167ef3057c90f7eR907-R910

Reply:
Update in [`3273142` (#61678)](https://github.com/pingcap/tidb/pull/61678/commits/327314292021cc0bcedb0c4eeae111e67c005925)